tests/unit/model-test.js

```diff
@@ -97,6 +97,52 @@ test("resetting a property on a record cause it to become clean again", function
   });
 });
 
+test("resetting a property to the current in-flight value causes it to become clean when the save completes", function(assert) {
+  assert.expect(4);
+
+  var person, saving, finishSaving;
+
+  env.adapter.updateRecord = function(store, type, snapshot) {
+    // Make sure the save is async
+    return new Ember.RSVP.Promise(function(resolve, reject) {
+      finishSaving = resolve;
+    });
+  };
+
+  run(function() {
+    store.push({
+      data: {
+        type: 'person',
+        id: '1',
+        attributes: {
+          name: 'Tom'
+        }
+      }
+    });
+    person = store.peekRecord('person', 1);
+    person.set('name', "Thomas");
+
+    saving = person.save();
+
+    assert.equal(person.get('name'), "Thomas");
+
+    person.set('name', 'Tomathy');
+    assert.equal(person.get('name'), "Tomathy");
+
+    person.set('name', 'Thomas');
+    assert.equal(person.get('name'), "Thomas");
```

Comment:
Without these 4 lines 129-133, this test passed without code changes. Setting `name` to 'Tomathy', then back to 'Thomas' put the record in a bad state